crypto/rand/rand_lib.c

```diff
@@ -670,13 +686,11 @@ unsigned char *rand_pool_add_begin(RAND_POOL *pool, size_t len)
         return NULL;
     }
 
-    if (pool->buffer == NULL) {
+    if (pool->buffer == NULL || len > pool->alloc_len - pool->len) {
         RANDerr(RAND_F_RAND_POOL_ADD_BEGIN, ERR_R_INTERNAL_ERROR);
         return NULL;
     }
 
-    if (!rand_pool_grow(pool, len))
-        return NULL;
```

Comment:
This assumes, that rand_pool_add_begin is only called for entropy buffers,
within the limits given by rand_pool_bytes_needed.
We sometimes use rand_pool_add also for entropy buffers, but within the limits of bytes_needed.
The buffer_grow in rand_pool_add is only needed for the case when additional data is added.
Question: is my assumption right that rand_pool_add_begin/end should not be used for additional data?

Reply:
> Question: is my assumption right that rand_pool_add_begin/end should not be used for additional data?

I can't currently answer your question in this generality without looking at the status-quo in the source code. When I implemented the pool, things were simpler, because the maximun allowed size was pre-allocated. But I can say that I never had such a limitation like get-entropy-only in mind.  I explicitly allowed setting min_entropy to zero. After all, it's just a generic pool of some data, so I did not see any reason why I should forbid using it for additional data.

Reply:
Yes, this is not in our current code base, but would might possibly be in the future.

Reply:
If it will be forbidden in the future, How do you want to guarantee that nobody ever forgets about it?

Reply:
I think given your above comment, it would be allowed but awkward programming style.
I have no problem with this rand_pool_grow in here, as long as the user
plays by the rule, he won't need to worry about error checking here.
Only when an adin pool adds data with this API it is necessary to check the
pointer to be non-zero.